### DISCUSSION TITLE
[BE] CI/CD 파이프라인 흐름 정리해보겠습니다.

### DISCUSSION THREAD

**Discussion by jujubebat** [darass-team/2021-darass#733] [General]
# CI/CD 파이프라인 과정
1. main에서 브랜치를 딴후 기능을 개발한다.
2. main 브랜치에 PR를 날린다.
   - 젠킨스에서 테스트를 포함한 빌드가 수행된다.
   - 소나큐브가 정적리포트를 분석한다.
   - 코드리뷰를 반영한다.
3. PR을 main에 머지한다.
   - 개발 서버로 자동 배포된다.
4. 개발 서버를 기준으로 QA를 진행한다.
5. 젠킨스에서 운영 서버로 수동 배포한다. 

# 수동 배포하는법
![image](https://user-images.githubusercontent.com/37281119/137344359-18d6c883-5a2b-4091-9f05-052532286d47.png)
- Build Now를 클릭한다. 

# 운영서버에서 버그가 발생할 경우 
- 서버에 접속해서 도커 레지스트리에 있는 이전 버전 이미지를 재배포한다.
- 매번 배포할때마다 도커 레지스트리에 배포 이미지를 남겨두고있다. 배포 히스토리를 관리하는 것이다. 이미지명 컨밴션은 아래와 같다.
  - darass/springboot-prod-server-{배포 당시 main 브랜치의 가장 마지막 커밋의 해쉬값}
- rebase로 main 브랜치의 hash 값이 바뀌지 않도록 주의한다. 

--- 

계속 논의한 것을 바탕으로 위와 같이 정리해봤습니다. 
제이온이 말해주신대로 추후 프론트엔드도 젠킨스로 일괄 처리하면 좋을 것 같습니다.
부족한 부분이나 개선사항 있으면 의견 남겨주세요. 😃

**Comment by zereight** [comment]
정리 감사합니당!
